### ISSUE TITLE
Feature requests

### ISSUE THREAD

**Issue by milesj** [ccbrown/iocraft#42]
Just going to note all my requests/ideas here:

### 1) Rename `Box` to `View`

Self-explanatory. Just collides with Rust native `Box`.

### 2) Support a `Handler` with a return value

I'm working on form/input components and I'm adding a validate function kind of prop, which works. https://github.com/moonrepo/starbase/blob/console/crates/console/src/components/input.rs#L13

However, I had to basically copy all of the `Handler` code to make it work, as I needed a return type of `Option<String>` (invalid with error message or valid). https://github.com/moonrepo/starbase/blob/console/crates/console/src/components/mod.rs#L29

Would be nice to have a built-in type for a handler with a return function. Adding it to the existing `Handler` may be too much, as we would then need to return `()` everywhere or something. So a new type maybe best.

### 3) Prop type that can be a `String` or `AnyElement`

I've found myself needing this pattern a lot just for convenience. Basically I want something like this:

```rust
struct ExampleProps {
  label: StringOrElement
}
```

So that when we're rendering, we can use a simple string (which would just render `Text` under the hood):

```
Example(label: &some_string_value)
```

Or we can customize with a specific element:

```
Example(label: element!(CustomLabel).into_any())
```

**Comment by milesj** [comment]
### 4) `render_loop` without ANSI

It looks like `render_loop` always renders with ANSI, even when in CI and non TTY environments, as seen here: https://github.com/moonrepo/proto/actions/runs/12342638090/job/34442711429?pr=671 Would be nice if this was turned off.

**Comment by ccbrown** [comment]
> ### 4) `render_loop` without ANSI
> It looks like `render_loop` always renders with ANSI, even when in CI and non TTY environments, as seen here: https://github.com/moonrepo/proto/actions/runs/12342638090/job/34442711429?pr=671 Would be nice if this was turned off.

The trick here is that `render_loop` can't function correctly without using ANSI sequences to erase prior output. A way to disable colors / text attributes would be reasonable though.

Also, for testing purposes, [`mock_terminal_render_loop`](https://docs.rs/iocraft/latest/iocraft/trait.ElementExt.html#tymethod.mock_terminal_render_loop) might meet your needs – it allows you to get output as a sequence of [`Canvas`](https://docs.rs/iocraft/latest/iocraft/prelude/struct.Canvas.html)s which you can then print or turn into a string without ANSI codes.

**Comment by milesj** [comment]
Yeah that makes sense. Was mainly referring to all the color related sequences.

**Comment by milesj** [comment]
### 5) Support non-primitive values in `use_state`

I have a situation where I want to use a `HashSet` in state, but I have no way of inserting a value into the set after it's created. `get` requires `Copy`, which sets aren't, while `read` is immutable. The `modify` function would work, but it's not public.

**Comment by ccbrown** [comment]
> ### 5) Support non-primitive values in `use_state`
> I have a situation where I want to use a `HashSet` in state, but I have no way of inserting a value into the set after it's created. `get` requires `Copy`, which sets aren't, while `read` is immutable. The `modify` function would work, but it's not public.

I just added a `write` function in https://github.com/ccbrown/iocraft/pull/45 (merged to main). Give that a try and let me know if that suits your needs. With that, you should be able to do something like `my_state.write().insert(foo)`.

Thanks as always for the excellent feedback!

**Comment by milesj** [comment]
Yeah that worked. Thanks for the quick turn around.

I did encounter another bug though. It looks like `ExtendWithElements::extend` collides with the native vec/map/set `extend`. May need to rename that method.

<img width="963" alt="Screenshot 2024-12-19 at 11 34 37 PM" src="https://github.com/user-attachments/assets/b4797255-d893-436c-b5bb-91ccea11ce90" />

**Comment by ccbrown** [comment]
> I did encounter another bug though. It looks like `ExtendWithElements::extend` collides with the native vec/map/set `extend`. May need to rename that method.

I might also want to exclude that trait from the prelude, but I've renamed the method for now and released it along with the `state.write` function in 0.5.1.

**Comment by milesj** [comment]
### 6) Support left/right arrows + caret in `TextInput`

This is a nice to have. Noticed while implementing inputs that `TextInput` doesn't allow left/right arrows, so you have to backspace to fix any issues. Would be nice to have this implemented natively. I tried to abstract around it, but since `TextInput` uses `push` internally, it won't work correctly. 

It would also be nice to render the "caret" of where the current position is. Probably with just a bg color.

If I have time, I can look into contributing this.

**Comment by ccbrown** [comment]
After experiencing more pain around number 1 here, I've decided it's time to bite the bullet and do the rename. It's better to do it sooner rather than later, and I don't know of any other upcoming breaking changes that it could be batched with.

This is merged to main ([2813e7992deec4aa68c1c6365bafb4617d200d2b](https://github.com/ccbrown/iocraft/commit/2813e7992deec4aa68c1c6365bafb4617d200d2b)) and will be released shortly as 0.6.0.

**Comment by milesj** [comment]
### 7) Allow a value to be returned from render

I'm sure this isn't possible because of Rust's type system, but it would be nice if the render functions could _return a value_ that was set by the component, probably with something like `system.exit_with_value(123)`. I just find the "use an &mut prop value to simulate on handler" to be extremely verbose, especially when code needs to be nested / has a bunch of conditionals.

For example, this code (from my app which has some abstractions):

```rust
let init = if options.yes {
    true
} else {
    let mut value = false;

    console
        .render_interactive(element! {
            Confirm(
                label: format!("Initialize moon into <path>{}</path>?", options.dir.display()),
                on_confirm: &mut value
            )
        })
        .await?;

    value
};
```

Would be a lot nicer written as:

```
let init = if options.yes {
    true
} else {
    console
        .render_interactive(element! {
            Confirm(
                label: format!("Initialize moon into <path>{}</path>?", options.dir.display()),
            )
        })
        .await
};
```

Perhaps the returned value type could be typed as a generic on `Hooks`? So `Hooks<usize>`?. Or maybe just have the value be `serde_json::Value` and ignore the type system. Just thinking out loud here.

**Comment by ccbrown** [comment]
Number 6 (left/right arrows in `TextInput`) is addressed by https://github.com/ccbrown/iocraft/pull/92. I still want to test it a bit more and see if I can make it a little better, but in the meantime, you can try the latest revision in that PR.

**Comment by ccbrown** [comment]
Text input cursors are now available in `main`. You can try it out with revision `34dd256f9f19628649d2c0a78aa0b1487e0c239c`. It'll be released soon in 0.7.7.
